src/subscriber.ts

```diff
@@ -14,497 +14,351 @@
-    this.flowControl = Object.assign(
-        {
-          maxBytes: os.freemem() * 0.2,
-          maxMessages: 100,
-        },
-        options.flowControl);
-    this.batching = Object.assign(
-        {
-          maxMilliseconds: 100,
-        },
-        options.batching);
-    this.flushTimeoutHandle_ = null;
-    this.leaseTimeoutHandle_ = null;
-    this.userClosed_ = false;
-    this.isOpen = false;
-    this.messageListeners = 0;
-    // As of right now we do not write any acks/modacks to the pull streams.
-    // But with allowing users to opt out of using streaming pulls altogether on
-    // the horizon, we may need to support this feature again in the near
-    // future.
-    this.writeToStreams_ = false;
-    this.listenForEvents_();
-  }
-  /*!
-   * Acks the provided message. If the connection pool is absent, it will be
-   * placed in an internal queue and sent out after 1 second or if the pool is
-   * re-opened before the timeout hits.
-   *
-   * @private
-   *
-   * @param {object} message The message object.
-   */
-  ack_(message) {
-    const breakLease = this.breakLease_.bind(this, message);
-    this.histogram.add(Date.now() - message.received);
-    if (this.writeToStreams_ && this.isConnected_()) {
-      this.acknowledge_(message.ackId, message.connectionId).then(breakLease);
-      return;
-    }
-    this.inventory_.ack.push(message.ackId);
-    this.setFlushTimeout_().then(breakLease);
+export class Message {
+  ackId: string;
+  attributes: {};
+  data: Buffer;
+  id: string;
+  publishTime: Date;
+  received: number;
+  private _length: number;
+  private _subscriber: Subscriber;
+  constructor(sub: Subscriber, {ackId, message}: ReceivedMessage) {
+    this.ackId = ackId;
+    this.attributes = message.attributes || {};
+    this.data = message.data;
+    this.id = message.messageId;
+    this.publishTime = Message.formatTimestamp(message.publishTime);
+    this.received = Date.now();
+    this._length = this.data.length;
+    this._subscriber = sub;
   }
-  /*!
-   * Sends an acknowledge request for the provided ack ids.
-   *
-   * @private
+  /**
+   * The length of the message data.
    *
-   * @param {string|string[]} ackIds The ack IDs to acknowledge.
-   * @param {string} [connId] Connection ID to send request on.
-   * @return {Promise}
+   * @type {number}
    */
-  acknowledge_(ackIds: string|string[], connId?: string) {
-    ackIds = arrify(ackIds);
-    const promises = chunk(ackIds, MAX_ACK_IDS_PER_REQUEST).map(ackIdChunk => {
-      if (this.writeToStreams_ && this.isConnected_()) {
-        return this.writeTo_(connId, {ackIds: ackIdChunk});
-      }
-      return promisify(this.request).call(this, {
-        client: 'SubscriberClient',
-        method: 'acknowledge',
-        reqOpts: {
-          subscription: this.name,
-          ackIds: ackIdChunk,
-        },
-      });
-    });
-    return Promise.all(promises).catch(err => {
-      this.emit('error', err);
-    });
+  get length() {
```

Comment:
The discussion is probably buried somewhere, but it is really important for us to preserve the original length of the message since we use it in combination with our `maxBytes` flow control option. This is just a precaution in case the user stomps on it for whatever reason.